packages/teleport/src/useAuditEvents.ts

```diff
@@ -15,48 +15,73 @@ limitations under the License.
 */
 
 import moment from 'moment';
-import React, { useEffect, useState, useMemo } from 'react';
-import { useAttempt } from 'shared/hooks';
+import { useEffect, useState, useMemo } from 'react';
+import useAttempt from 'shared/hooks/useAttemptNext';
 import Ctx from 'teleport/teleportContext';
+import { Event } from 'teleport/services/audit';
 
-export default function useEvents(ctx: Ctx, clusterId: string) {
+export default function useAuditEvents(
+  ctx: Ctx,
+  clusterId: string,
+  limit?: number
+) {
   const rangeOptions = useMemo(() => getRangeOptions(), []);
-  const [searchValue, setSearchValue] = React.useState('');
-  const [range, setRange] = useState(rangeOptions[0]);
-  const [attempt, attemptActions] = useAttempt({ isProcessing: true });
-  const [results, setResults] = useState({
-    events: [],
-    overflow: false,
-  });
-
-  function onFetch({ from, to }: Range) {
-    attemptActions.do(() => {
-      return ctx.auditService
-        .fetchEvents(clusterId, { start: from, end: to })
-        .then(setResults);
-    });
-  }
+  const [searchValue, setSearchValue] = useState('');
+  const [range, setRange] = useState<EventRange>(rangeOptions[0]);
+  const { attempt, setAttempt, run } = useAttempt('processing');
+  const [events, setEvents] = useState<Event[]>([]);
+  const [startKey, setStartKey] = useState('');
+  const [fetchStatus, setFetchStatus] = useState<FetchStatus>('');
 
   useEffect(() => {
-    onFetch(range);
+    fetch();
   }, [clusterId, range]);
 
+  // fetchMore gets events from last position from
+  // last fetch, indicated by startKey. The response is
+  // appended to existing events list.
+  function fetchMore() {
+    setFetchStatus('loading');
+    ctx.auditService
+      .fetchEvents(clusterId, { ...range, startKey })
+      .then(res => {
+        setEvents([...events, ...res.events]);
+        setStartKey(res.startKey);
+        setFetchStatus(res.startKey ? '' : 'disabled');
```

Comment:
it will trigger multiple re-renders. This might be fine if the table was hidden while fetching but clicking on `fetchMore` does not hide so it will re-render a table with thousands of data 3 times in a row. 

Instead do:

```tsx
const [results, setResults] = useState({
    events: [],
    fetchStartKey: xxx,
    fetchStatus: false,
});

ctx.auditService
      .fetchEvents(clusterId, { ...range, startKey })
      .then(res => {
        setState({
          events: [...events, ...res.events],
          fetchStatus: ...
          fetchStartKey: ...
        }));
```

Reply:
revised